golang/x/relay/keeper/querier.go

```diff
@@ -34,48 +73,242 @@ func queryIsAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keep
+	ancestor, ancestorErr := hash256DigestFromHex(path[1])
+	if ancestorErr != nil {
+		return []byte{}, ancestorErr
 	}
 
-	// TODO: check that the path is this many items long, error if not
-	limit := uint32(25) // TODO: parse from path, use 240 as default if not in path
+	var limit uint32
+	if len(path) == 2 {
+		limit = 15
+	} else {
+		num, convErr := strconv.ParseUint(path[2], 0, 10)
+		if convErr != nil {
+			return []byte{}, types.ErrExternal(types.DefaultCodespace, convErr)
+		}
+		limit = uint32(num)
+	}
 
 	// This calls the keeper with the parsed arguments, and gets an answer
-	result := keeper.IsAncestor(ctx, ancestorDigestLE, digestLE, limit)
+	result := keeper.IsAncestor(ctx, ancestor, digestLE, uint32(limit))
 
 	// Now we format the answer as a response
 	response := types.QueryResIsAncestor{
 		Digest:              digestLE,
-		ProspectiveAncestor: ancestorDigestLE,
-		IsAncestor:          result,
+		ProspectiveAncestor: ancestor,
+		Res:                 result,
+	}
+
+	// And we serialize that response as JSON
+	res, marshalErr := codec.MarshalJSONIndent(keeper.cdc, response)
+	if marshalErr != nil {
+		return []byte{}, types.ErrMarshalJSON(types.DefaultCodespace)
+	}
+	return res, nil
+}
+
+func queryGetRelayGenesis(ctx sdk.Context, req abci.RequestQuery, keeper Keeper) (res []byte, err sdk.Error) {
+	// This calls the keeper and gets an answer
+	result, err := keeper.GetRelayGenesis(ctx)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	// Now we format the answer as a response
+	response := types.QueryResGetRelayGenesis{
+		Res: result,
+	}
+
+	// And we serialize that response as JSON
+	res, marshalErr := codec.MarshalJSONIndent(keeper.cdc, response)
+	if marshalErr != nil {
+		return []byte{}, types.ErrMarshalJSON(types.DefaultCodespace)
+	}
+	return res, nil
+}
+
+func queryGetLastReorgLCA(ctx sdk.Context, req abci.RequestQuery, keeper Keeper) (res []byte, err sdk.Error) {
+	// This calls the keeper and gets an answer
+	result, err := keeper.GetRelayGenesis(ctx)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	// Now we format the answer as a response
+	response := types.QueryResGetLastReorgLCA{
+		Res: result,
+	}
+
+	// And we serialize that response as JSON
+	res, marshalErr := codec.MarshalJSONIndent(keeper.cdc, response)
+	if marshalErr != nil {
+		return []byte{}, types.ErrMarshalJSON(types.DefaultCodespace)
+	}
+	return res, nil
+}
+
+func queryFindAncestor(ctx sdk.Context, path []string, req abci.RequestQuery, keeper Keeper) (res []byte, err sdk.Error) {
+	// check that the path is this many items long, error if not
+	if len(path) > 2 {
+		return []byte{}, types.ErrTooManyArguments(types.DefaultCodespace)
+	} else if len(path) < 2 {
+		return []byte{}, types.ErrNotEnoughArguments(types.DefaultCodespace)
+	}
+
+	digestLE, digestErr := hash256DigestFromHex(path[0])
+	if digestErr != nil {
+		return []byte{}, digestErr
+	}
+
+	offset, convErr := strconv.ParseUint(path[1], 0, 10)
```

Comment:
the 3rd argument of `ParseUint` is the number of bits the number fits in. it should be `32` for all of these

Reply:
Oh, fixed